File: task2/memory_sim.py
import sys

from networkx import hits
PAGE_SIZE = 4 # step 1 configuration variable[cite: 1, 2]
TOTAL_FRAMES = 3
#core helpers
def get_page_number(virtual_address):
    # simple integer division to drop the offset bytes
    """converts a raw memory address into its corresponding page number."""
    return virtual_address // PAGE_SIZE

#grid visualization formatter
def print_memory_grid(algorithm_name, current_frames, status, page_num):
    """prints a clear, human-readable grid layout of physical frames."""
    frame_display = []
    for i in range(TOTAL_FRAMES):
        if i < len(current_frames):
                    frame_display.append(f"[{current_frames[i]}]")
        else:
                    frame_display.append("[ ]")  # empty slot representation

    grid_string = " | ".join(frame_display)
    print(
                f"[{algorithm_name}] Page {page_num} -> {status.upper().ljust(5)} | Frames: {grid_string}"
     )
    
    #algorithm simulation interfaces
def run_fifo_simulation(address_stream):
    """simulates first-in-first-out page replacement logic."""
    print(f"\n--- starting fifo simulation (frames available: {TOTAL_FRAMES}) ---")
    frames = []#for tracking pages physically resident in memory
    hits=0
    faults = 0
    
    for addr in address_stream:
        page = get_page_number(addr)
        
        if page not in frames:
            faults += 1
            # checkimg if memory grid has open capacity
            if len(frames) < TOTAL_FRAMES:
                frames.append(page)
            else:
                # first in is at index 0, pop it out[cite: 1]
                frames.pop(0)
                frames.append(page)
            print_memory_grid("FIFO", frames, "fault", page)
        else:
            hits += 1
            print_memory_grid("FIFO", frames, "hit", page)
     #ratio calculations for performance metrics       
    total_ops = len(address_stream)
    hit_ratio = (hits / total_ops) * 100
    miss_ratio = (faults / total_ops) * 100

    print(f"fifo results -> hits: {hits}, faults: {faults}")
    print(
        f"performance ratios -> hit ratio: {hit_ratio:.1f}%, miss ratio: {miss_ratio:.1f}%\n"
    )
    return faults

#lru algorithm interface
def run_lru_simulation(address_stream):
    """simulates least-recently-used page replacement logic."""
    print(f"\n--- starting lru simulation (frames available: {TOTAL_FRAMES}) ---")
    frames = []
    access_history = {} # dictionary tracker to keep logical clock ticks
    hits=0
    faults = 0
    logical_time = 0 # increments on every page reference step
    
    for addr in address_stream:
        logical_time += 1
        page = get_page_number(addr)
        
        # update time rank on every hit or miss access alike
        access_history[page] = logical_time
        
        if page in frames:
            hits += 1
            print_memory_grid("LRU", frames, "hit", page)
        else:
            faults += 1
            if len(frames) < TOTAL_FRAMES:
                frames.append(page)
            else:
                # search the active grid to track down the minimum used frame slot
                oldest_page = frames[0]
                min_time = access_history[oldest_page]
                
                for f in frames:
                    if access_history[f] < min_time:
                        min_time = access_history[f]
                        oldest_page = f
                        
                # boot out the unaccessed page context link
                frames.remove(oldest_page)
                frames.append(page)
            print_memory_grid("LRU", frames, "fault", page)
      #ratio calculation integration for performance metrics      
    total_ops = len(address_stream)
    hit_ratio = (hits / total_ops) * 100
    miss_ratio = (faults / total_ops) * 100

    print(f"lru results -> hits: {hits}, faults: {faults}")
    print(
        f"performance ratios -> hit ratio: {hit_ratio:.1f}%, miss ratio: {miss_ratio:.1f}%\n"
    )
    return faults

if __name__ == "__main__":
    # quick sanity verification test loop
    addresses = [2, 3, 5, 11]
    for addr in addresses:
        print("address: " + str(addr) + " maps directly to page: " + str(get_page_number(addr)))
        
    print("\n--- running actual trace simulation ---")
    virtual_addresses = [2, 3, 5, 11, 1, 14, 7, 19]

    fifo_faults = run_fifo_simulation(virtual_addresses)
    lru_faults = run_lru_simulation(virtual_addresses)
    
    print("comparison overview ---->")
    print(f"total system page faults generated -> fifo: {fifo_faults} | lru: {lru_faults}")

File: task2/memory_sim.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define PAGE_SIZE 4    // step 1 configuration variable
#define TOTAL_FRAMES 3

// core helpers
// converts a raw memory address into its corresponding page number.
int get_page_number(int virtual_address) {
    // simple integer division to drop the offset bytes
    return virtual_address / PAGE_SIZE;
}

// grid visualization formatter
// prints a clear, human-readable grid layout of physical frames.
void print_memory_grid(const char *algorithm_name, int *frames, int frame_count, const char *status, int page_num) {
    char status_upper[16];
    int j = 0;
    while (status[j] != '\0' && j < 15) {
        if (status[j] >= 'a' && status[j] <= 'z') {
            status_upper[j] = status[j] - 32;
        } else {
            status_upper[j] = status[j];
        }
        j++;
    }
    status_upper[j] = '\0';

    printf("[%s] Page %d -> %-5s | Frames: ", algorithm_name, page_num, status_upper);
    for (int i = 0; i < TOTAL_FRAMES; i++) {
        if (i < frame_count) {
            printf("[%d]", frames[i]);
        } else {
            printf("[ ]"); // empty slot representation
        }
        if (i < TOTAL_FRAMES - 1) {
            printf(" | ");
        }
    }
    printf("\n");
}

// algorithm simulation interfaces
// simulates first-in-first-out page replacement logic.
int run_fifo_simulation(int *address_stream, int stream_len) {
    printf("\n--- starting fifo simulation (frames available: %d) ---\n", TOTAL_FRAMES);
    int frames[TOTAL_FRAMES]; // for tracking pages physically resident in memory
    int frame_count = 0;
    int hits = 0;
    int faults = 0;

    for (int idx = 0; idx < stream_len; idx++) {
        int addr = address_stream[idx];
        int page = get_page_number(addr);

        int found = 0;
        for (int f = 0; f < frame_count; f++) {
            if (frames[f] == page) {
                found = 1;
                break;
            }
        }

        if (!found) {
            faults++;
            // checkimg if memory grid has open capacity
            if (frame_count < TOTAL_FRAMES) {
                frames[frame_count++] = page;
            } else {
                // first in is at index 0, pop it out
                for (int f = 0; f < TOTAL_FRAMES - 1; f++) {
                    frames[f] = frames[f + 1];
                }
                frames[TOTAL_FRAMES - 1] = page;
            }
            print_memory_grid("FIFO", frames, frame_count, "fault", page);
        } else {
            hits++;
            print_memory_grid("FIFO", frames, frame_count, "hit", page);
        }
    }

    // ratio calculations for performance metrics
    double hit_ratio = ((double)hits / stream_len) * 100.0;
    double miss_ratio = ((double)faults / stream_len) * 100.0;

    printf("fifo results -> hits: %d, faults: %d\n", hits, faults);
    printf("performance ratios -> hit ratio: %.1f%%, miss ratio: %.1f%%\n\n", hit_ratio, miss_ratio);
    return faults;
}

// lru algorithm interface
// simulates least-recently-used page replacement logic.
int run_lru_simulation(int *address_stream, int stream_len) {
    printf("\n--- starting lru simulation (frames available: %d) ---\n", TOTAL_FRAMES);
    int frames[TOTAL_FRAMES];
    int access_history[1000] = {0}; // array tracker to keep logical clock ticks
    int frame_count = 0;
    int hits = 0;
    int faults = 0;
    int logical_time = 0; // increments on every page reference step

    for (int idx = 0; idx < stream_len; idx++) {
        logical_time++;
        int addr = address_stream[idx];
        int page = get_page_number(addr);

        // update time rank on every hit or miss access alike
        access_history[page] = logical_time;

        int found = 0;
        for (int f = 0; f < frame_count; f++) {
            if (frames[f] == page) {
                found = 1;
                break;
            }
        }

        if (found) {
            hits++;
            print_memory_grid("LRU", frames, frame_count, "hit", page);
        } else {
            faults++;
            if (frame_count < TOTAL_FRAMES) {
                frames[frame_count++] = page;
            } else {
                // search the active grid to track down the minimum used frame slot
                int oldest_page = frames[0];
                int min_time = access_history[oldest_page];
                int oldest_idx = 0;

                for (int f = 0; f < frame_count; f++) {
                    if (access_history[frames[f]] < min_time) {
                        min_time = access_history[frames[f]];
                        oldest_page = frames[f];
                        oldest_idx = f;
                    }
                }

                // boot out the unaccessed page context link
                for (int f = oldest_idx; f < frame_count - 1; f++) {
                    frames[f] = frames[f + 1];
                }
                frames[frame_count - 1] = page;
            }
            print_memory_grid("LRU", frames, frame_count, "fault", page);
        }
    }

    // ratio calculation integration for performance metrics
    double hit_ratio = ((double)hits / stream_len) * 100.0;
    double miss_ratio = ((double)faults / stream_len) * 100.0;

    printf("lru results -> hits: %d, faults: %d\n", hits, faults);
    printf("performance ratios -> hit ratio: %.1f%%, miss ratio: %.1f%%\n\n", hit_ratio, miss_ratio);
    return faults;
}

int main(void) {
    // quick sanity verification test loop
    int addresses[] = {2, 3, 5, 11};
    int num_addresses = sizeof(addresses) / sizeof(addresses[0]);
    for (int i = 0; i < num_addresses; i++) {
        printf("address: %d maps directly to page: %d\n", addresses[i], get_page_number(addresses[i]));
    }

    printf("\n--- running actual trace simulation ---\n");
    int virtual_addresses[] = {2, 3, 5, 11, 1, 14, 7, 19};
    int stream_len = sizeof(virtual_addresses) / sizeof(virtual_addresses[0]);

    int fifo_faults = run_fifo_simulation(virtual_addresses, stream_len);
    int lru_faults = run_lru_simulation(virtual_addresses, stream_len);

    printf("comparison overview ---->\n");
    printf("total system page faults generated -> fifo: %d | lru: %d\n", fifo_faults, lru_faults);
    return 0;
}
